File: sdk/android/android/src/main/kotlin/dev/chill/android/Replay.kt
package dev.chill.android

import android.graphics.Rect
import android.view.View
import android.view.ViewGroup
import dev.chill.core.ChillRuntime

public data class ReplayNode(val role: String, val resourceId: String?, val frame: Rect, val masked: Boolean = true, val children: List<ReplayNode> = emptyList())
public object ViewReplayCapture {
    public fun snapshot(root: View): ReplayNode { val location = IntArray(2); root.getLocationOnScreen(location); val children = if (root is ViewGroup) (0 until root.childCount).map { snapshot(root.getChildAt(it)) } else emptyList(); val id = if (root.id == View.NO_ID) null else runCatching { root.resources.getResourceEntryName(root.id) }.getOrNull(); return ReplayNode(role(root), id, Rect(location[0], location[1], location[0] + root.width, location[1] + root.height), children = children) }
}

public class ReplayRecorder(private val runtime: ChillRuntime, private val maximumChunks: Int = 1_000) {
    private var sequence = 0; private var dropped = 0; private val started = System.nanoTime()
    public fun snapshot(root: View) { capture("snapshot", mapOf("tree" to ViewReplayCapture.snapshot(root).toString())) }
    public fun gesture(x: Float, y: Float, type: String) { capture("gesture", mapOf("x" to x.toInt(), "y" to y.toInt(), "gesture" to type)) }
    public fun viewport(width: Int, height: Int) { capture("viewport", mapOf("width" to width, "height" to height)) }
    public fun finish() { if (dropped > 0) runtime.replay(mapOf("replay_id" to runtime.replayId, "sequence" to sequence, "type" to "gap", "dropped_chunks" to dropped, "reason" to "capacity", "masking" to "source")) }
    private fun capture(type: String, payload: Map<String, Any>) { if (!runtime.shouldReplay()) return; if (sequence >= maximumChunks) { dropped++; return }; runtime.replay(payload + mapOf("replay_id" to runtime.replayId, "sequence" to sequence++, "monotonic_offset_nano" to (System.nanoTime() - started).toString(), "type" to type, "masking" to "source")) }
}

File: sdk/web/src/replay.ts
import type { ChillBrowser } from "./runtime.js";

interface ReplayNode { tag: string; id?: string; role?: string; masked: true; children: ReplayNode[]; }
function serialize(element: Element): ReplayNode {
  const node: ReplayNode = { tag: element.tagName.toLowerCase(), masked: true, children: [...element.children].map(serialize) };
  const id = (element as HTMLElement).dataset.chillId; if (id) node.id = id;
  const role = element.getAttribute("role"); if (role) node.role = role;
  return node;
}

export function startReplay(client: ChillBrowser, root: HTMLElement = document.documentElement): () => void {
  if (!client.shouldReplay()) return () => undefined;
  let sequence = 0; let dropped = 0; const maximumChunks = 1_000; const started = performance.now();
  const capture = (type: string, payload: Record<string, unknown>) => { if (sequence >= maximumChunks) { dropped += 1; return; } client.replay({ replay_id: client.replayId, sequence: sequence++, monotonic_offset_nano: `${Math.trunc((performance.now() - started) * 1_000_000)}`, type, masking: "source", ...payload }); };
  capture("snapshot", { tree: serialize(root), viewport: { width: innerWidth, height: innerHeight } });
  const observer = new MutationObserver(records => capture("mutation", { mutations: records.map(record => ({ type: record.type, target: record.target instanceof Element ? record.target.tagName.toLowerCase() : "text", added_count: record.addedNodes.length, removed_count: record.removedNodes.length })) }));
  observer.observe(root, { subtree: true, childList: true, attributes: true });
  const pointer = (event: PointerEvent) => capture("gesture", { gesture: event.type, x: Math.round(event.clientX), y: Math.round(event.clientY) });
  const scroll = () => capture("scroll", { x: Math.round(scrollX), y: Math.round(scrollY) });
  root.addEventListener("pointerdown", pointer, true); addEventListener("scroll", scroll, { capture: true, passive: true });
  return () => { observer.disconnect(); root.removeEventListener("pointerdown", pointer, true); removeEventListener("scroll", scroll, true); if (dropped > 0) client.replay({ replay_id: client.replayId, sequence, type: "gap", dropped_chunks: dropped, reason: "capacity", masking: "source" }); };
}
